File: web/dsp/prng.js
/**
 * xoshiro128** - a small deterministic generator, implemented identically here
 * and in Python (natvox/dsp/prng.py).
 *
 * The engine needs randomness in two places: the jitter that keeps unvoiced
 * grain spacing from stamping a buzz onto fricatives, and the noise for the
 * breath mix. Using each language's native generator would make the two
 * implementations diverge on any signal containing consonants, which would
 * destroy the only cheap way to verify the port did not silently degrade:
 * feeding both the same audio and diffing the samples.
 *
 * All arithmetic is 32-bit, which JavaScript does natively.
 */
export class Prng {
  /** @param {number} seed any 32-bit integer */
  constructor(seed = 0x5eed) {
    // SplitMix-style seeding so that nearby seeds do not produce similar streams.
    let s = seed >>> 0;
    this.s = new Uint32Array(4);
    for (let i = 0; i < 4; i++) {
      s = (s + 0x9e3779b9) >>> 0;
      let z = s;
      z = Math.imul(z ^ (z >>> 16), 0x21f0aaad) >>> 0;
      z = Math.imul(z ^ (z >>> 15), 0x735a2d97) >>> 0;
      this.s[i] = (z ^ (z >>> 15)) >>> 0;
    }
    this._spare = null;
  }

  /** Next raw 32-bit value. */
  next() {
    const s = this.s;
    const result = (Math.imul(rotl(Math.imul(s[1], 5) >>> 0, 7), 9) >>> 0);
    const t = (s[1] << 9) >>> 0;
    s[2] = (s[2] ^ s[0]) >>> 0;
    s[3] = (s[3] ^ s[1]) >>> 0;
    s[1] = (s[1] ^ s[2]) >>> 0;
    s[0] = (s[0] ^ s[3]) >>> 0;
    s[2] = (s[2] ^ t) >>> 0;
    s[3] = rotl(s[3], 11);
    return result;
  }

  /** Uniform in [0, 1) with 24 bits of resolution. */
  uniform() {
    return (this.next() >>> 8) * (1 / 16777216);
  }

  /** Uniform in [lo, hi). */
  range(lo, hi) {
    return lo + (hi - lo) * this.uniform();
  }

  /** Standard normal, Box-Muller with the second value cached. */
  normal() {
    if (this._spare !== null) {
      const v = this._spare;
      this._spare = null;
      return v;
    }
    // Avoid log(0); uniform() can return exactly 0.
    let u1 = this.uniform();
    if (u1 < 1e-12) u1 = 1e-12;
    const u2 = this.uniform();
    const r = Math.sqrt(-2 * Math.log(u1));
    const theta = 2 * Math.PI * u2;
    this._spare = r * Math.sin(theta);
    return r * Math.cos(theta);
  }
}

function rotl(x, k) {
  return (((x << k) | (x >>> (32 - k))) >>> 0);
}

File: natvox/dsp/prng.py
"""xoshiro128** - a small deterministic generator, matched byte for byte by the
JavaScript port in ``web/dsp/prng.js``.

The engine needs randomness in two places: the jitter that keeps unvoiced grain
spacing from stamping a buzz onto fricatives, and the noise for the breath mix.
Using numpy's generator here and something else in JavaScript would make the two
implementations diverge on any signal containing consonants, which would destroy
the cheapest way to verify the port did not silently degrade quality: feed both
the same audio and diff the samples.

Quality beyond "uncorrelated" is not needed -- this seeds inaudible jitter and a
noise bed, not cryptography -- so a generator that is trivially portable wins
over a better one that is not.
"""
from __future__ import annotations

import math

import numpy as np

MASK32 = 0xFFFFFFFF


def _rotl(x: int, k: int) -> int:
    return ((x << k) | (x >> (32 - k))) & MASK32


class Prng:
    """Deterministic 32-bit generator; see module docstring."""

    def __init__(self, seed: int = 0x5EED) -> None:
        s = seed & MASK32
        state = []
        for _ in range(4):
            s = (s + 0x9E3779B9) & MASK32
            z = s
            z = ((z ^ (z >> 16)) * 0x21F0AAAD) & MASK32
            z = ((z ^ (z >> 15)) * 0x735A2D97) & MASK32
            state.append((z ^ (z >> 15)) & MASK32)
        self._s = state
        self._spare: float | None = None

    def next(self) -> int:
        s = self._s
        result = (_rotl((s[1] * 5) & MASK32, 7) * 9) & MASK32
        t = (s[1] << 9) & MASK32
        s[2] ^= s[0]
        s[3] ^= s[1]
        s[1] ^= s[2]
        s[0] ^= s[3]
        s[2] ^= t
        s[3] = _rotl(s[3], 11)
        return result

    def uniform(self) -> float:
        return (self.next() >> 8) * (1.0 / 16777216.0)

    def range(self, lo: float, hi: float) -> float:
        return lo + (hi - lo) * self.uniform()

    def normal(self) -> float:
        if self._spare is not None:
            value, self._spare = self._spare, None
            return value
        u1 = self.uniform()
        if u1 < 1e-12:
            u1 = 1e-12
        u2 = self.uniform()
        r = math.sqrt(-2.0 * math.log(u1))
        theta = 2.0 * math.pi * u2
        self._spare = r * math.sin(theta)
        return r * math.cos(theta)

    def normals(self, count: int) -> np.ndarray:
        """``count`` standard normals as an array."""
        return np.fromiter((self.normal() for _ in range(count)),
                           dtype=np.float64, count=count)
